File: sorting-searching/SortingSearching.cpp
#include <algorithm>
#include <iostream>
#include <vector>

// --- Manual algorithms ---

std::vector<int> bubbleSort(std::vector<int> arr) {
    int n = static_cast<int>(arr.size());
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n - i - 1; ++j) {
            if (arr[j] > arr[j + 1]) {
                std::swap(arr[j], arr[j + 1]);
            }
        }
    }
    return arr;
}

std::vector<int> merge(const std::vector<int>& left, const std::vector<int>& right) {
    std::vector<int> result;
    size_t i = 0, j = 0;
    while (i < left.size() && j < right.size()) {
        if (left[i] <= right[j]) result.push_back(left[i++]);
        else result.push_back(right[j++]);
    }
    while (i < left.size()) result.push_back(left[i++]);
    while (j < right.size()) result.push_back(right[j++]);
    return result;
}

std::vector<int> mergeSort(std::vector<int> arr) {
    if (arr.size() <= 1) return arr;
    size_t mid = arr.size() / 2;
    auto left = mergeSort({arr.begin(), arr.begin() + mid});
    auto right = mergeSort({arr.begin() + mid, arr.end()});
    return merge(left, right);
}

int binarySearch(const std::vector<int>& arr, int target) {
    int lo = 0, hi = static_cast<int>(arr.size()) - 1;
    while (lo <= hi) {
        int mid = (lo + hi) / 2;
        if (arr[mid] == target) return mid;
        else if (arr[mid] < target) lo = mid + 1;
        else hi = mid - 1;
    }
    return -1;
}

void print(const std::string& label, const std::vector<int>& arr) {
    std::cout << label << " [";
    for (size_t i = 0; i < arr.size(); ++i) {
        std::cout << arr[i] << (i + 1 < arr.size() ? ", " : "");
    }
    std::cout << "]\n";
}

int main() {
    std::cout << "Language: C++\n";
    std::vector<int> unsorted = {64, 34, 25, 12, 22, 11, 90};
    print("Original:        ", unsorted);

    // Manual sorts
    print("Bubble sort:     ", bubbleSort(unsorted));
    print("Merge sort:      ", mergeSort(unsorted));

    // Built-in sort (introsort)
    std::vector<int> builtIn = unsorted;
    std::sort(builtIn.begin(), builtIn.end());
    print("std::sort():     ", builtIn);

    // Manual binary search
    std::vector<int> sorted = mergeSort(unsorted);
    std::cout << "Manual binary search for 25: index " << binarySearch(sorted, 25) << "\n";

    // Built-in binary search (lower_bound)
    auto it = std::lower_bound(sorted.begin(), sorted.end(), 25);
    int idx = static_cast<int>(it - sorted.begin());
    std::cout << "std::lower_bound for 25:     index " << idx << "\n";

    // std::binary_search (returns bool, not index)
    bool found = std::binary_search(sorted.begin(), sorted.end(), 25);
    std::cout << "std::binary_search found 25: " << (found ? "true" : "false") << "\n";

    return 0;
}

File: sorting-searching/sorting_searching.py
from bisect import bisect_left


# --- Manual algorithms ---

def bubble_sort(arr: list[int]) -> list[int]:
    """O(n^2) comparison sort."""
    result = arr[:]
    n = len(result)
    for i in range(n):
        for j in range(0, n - i - 1):
            if result[j] > result[j + 1]:
                result[j], result[j + 1] = result[j + 1], result[j]
    return result


def merge_sort(arr: list[int]) -> list[int]:
    """O(n log n) divide-and-conquer sort."""
    if len(arr) <= 1:
        return arr[:]
    mid = len(arr) // 2
    left = merge_sort(arr[:mid])
    right = merge_sort(arr[mid:])
    return _merge(left, right)


def _merge(left: list[int], right: list[int]) -> list[int]:
    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result


def binary_search(arr: list[int], target: int) -> int:
    """Returns index of target in sorted arr, or -1 if not found."""
    lo, hi = 0, len(arr) - 1
    while lo <= hi:
        mid = (lo + hi) // 2
        if arr[mid] == target:
            return mid
        elif arr[mid] < target:
            lo = mid + 1
        else:
            hi = mid - 1
    return -1


def main() -> None:
    print("Language: Python")
    unsorted = [64, 34, 25, 12, 22, 11, 90]
    print("Original:         ", unsorted)

    # Manual sorts
    print("Bubble sort:      ", bubble_sort(unsorted))
    print("Merge sort:       ", merge_sort(unsorted))

    # Built-in sort (Timsort)
    built_in = sorted(unsorted)
    print("Built-in sorted() :", built_in)

    # Manual binary search
    sorted_arr = sorted(unsorted)
    idx = binary_search(sorted_arr, 25)
    print(f"Manual binary search for 25: index {idx}")

    # Built-in binary search (bisect)
    bi_idx = bisect_left(sorted_arr, 25)
    print(f"bisect_left for 25:          index {bi_idx}")

    # Built-in search in list
    print(f"list.index() for 25:         index {sorted_arr.index(25)}")


if __name__ == "__main__":
    main()
